### PR TITLE
made protocols and providers async (GDEV-716)

### PR DESCRIPTION
**Pull request by KerstenBreuer**
Description:
```
* Made eventpub protocol + Kafka provider async

* tested ascii validation of protocols separately

* Made eventsub protocol + Kafka provider asyn

* update github action deps

* adapt example
```

### CODE REVIEW

File: tests/unit/test_eventpub.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+)
+async def test_ascii_val(type_, key, topic, exception):
+    """Tests the ASCII validation logic included in the EventPublisherProtocol."""
+    payload = {"test_content": "Hello World"}
+
+    # create event publisher:
+    event_publisher = FakePublisher()
+
+    # publish event using the provider:
+    with (pytest.raises(exception) if exception else nullcontext()):  # type: ignore
```
Comment (Cito) [inline tests/unit/test_eventpub.py:70]: No outer parens needed here.

File: tests/unit/test_eventsub.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+)
+async def test_ascii_val(type_, topic, exception):
+    """Tests the ASCII validation logic included in the EventSubscriberProtocol."""
+    payload = {"test_content": "Hello World"}
+
+    # create event publisher:
+    event_submitter = FakeSubscriber()
+
+    # publish event using the provider:
+    with (pytest.raises(exception) if exception else nullcontext()):  # type: ignore
```
Comment (Cito) [inline tests/unit/test_eventsub.py:62]: No outer braces needed

File: tests/unit/test_in_mem_pub.py [hunk 3]
```diff
@@ -16,44 +16,22 @@
...
-            "test_ßtopic",  # non ascii
-            NonAsciiStrError,
-        ),
-    ],
-)
-def test_in_mem_publisher(type_, key, topic, exception):
+
+
+@pytest.mark.asyncio
+async def test_in_mem_publisher():
```
Comment (Cito) [inline tests/unit/test_in_mem_pub.py:29]: See same comment for KafkaPublisher. The error tests may still make sense to check that super() is used by the InMemPublisher.

**Review by Cito — APPROVED** [review]
Nice. Added some minor remarks, but feel free to ignore them.

File: setup.cfg [hunk 4]
```diff
@@ -49,7 +49,7 @@ python_requires = >= 3.9
 
 [options.extras_require]
 akafka =
-    kafka-python>=2.0.2
+    aiokafka>=0.7.2
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply setup.cfg:52]: Good idea.

File: tests/unit/test_akafka.py [hunk 5]
```diff
@@ -16,67 +16,38 @@
...
-)
-def test_kafka_event_publisher(
-    type_: str,
-    key: str,
-    topic: str,
-    expected_headers: list[tuple[str, bytes]],
-    exception: Optional[type[Exception]],
-):
+@pytest.mark.asyncio
+async def test_kafka_event_publisher():
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply tests/unit/test_akafka.py:31]: This is a good point. I was thinking a bit more and probably it's either way not a good design to rely on the implementer (in case of the publishing protocol that is us, but in case of the subscription protocol that is whomever is using our library) to call super().

Maybe it is cleaner to have the validation logic in a non-abstract method which is calling another private method. This private method is marked as abstract and must be implemented by the developer that subclasses from that abstract class.  
Like this we would nicely separate the concerns of users and implementors in two separate methods.

Please have a look at this commit, as an example: https://github.com/ghga-de/hexkit/pull/12/commits/d599600b66ff7838ed85da4e231cc99af3a9fe0e

File: tests/unit/test_akafka.py [hunk 6]
```diff
@@ -86,33 +57,25 @@ def test_kafka_event_publisher(
...
             key=key,
-            headers=expected_headers,
+            topic=topic,
         )
-        producer.flush.assert_called_once()
-    # check if producer was closed correctly:
-    producer.close.assert_called_once()
+
+    # check if producer was correctly used:
+    producer.start.assert_awaited_once()
```
[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply tests/unit/test_akafka.py]: Yes, in deed, that would be better.

[the remark this answers is not in the document]
Comment (KerstenBreuer) [reply tests/unit/test_akafka.py]: ✅

File: tests/unit/test_eventpub.py (hunk 1 above)
Comment (KerstenBreuer) [reply tests/unit/test_eventpub.py:70]: Right

Comment (KerstenBreuer) [reply tests/unit/test_eventpub.py]: ✅

File: tests/unit/test_eventsub.py (hunk 2 above)
Comment (KerstenBreuer) [reply tests/unit/test_eventsub.py]: ✅

File: tests/unit/test_akafka.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Cito) [reply tests/unit/test_akafka.py:31]: Yes, that's a nice solution. +1

File: hexkit/protocols/eventsub.py [hunk 7]
```diff
@@ -55,4 +54,18 @@ def consume(self, *, payload: JsonObject, type_: Ascii, topic: Ascii) -> None:
...
         """
         check_ascii(type_, topic)
+        await self._consume_validated(payload=payload, type_=type_, topic=topic)
+
+    @abstractmethod
+    async def _consume_validated(
+        self, *, payload: JsonObject, type_: Ascii, topic: Ascii
+    ) -> None:
+        """
+        Receive a pre-validated event of interest and process it according to its type.
```
Comment (Cito) [inline hexkit/protocols/eventsub.py:64]: This description is a bit misleading, you might think that the event itself is or was validated.

"Receive and process an event with already validated topic and type."?

File: hexkit/protocols/eventpub.py [hunk 8]
```diff
@@ -38,4 +37,20 @@ def publish(
             topic (str): The event type. ASCII characters only.
         """
         check_ascii(type_, key, topic)
+        await self._publish_validated(
```
Comment (Cito) [inline hexkit/protocols/eventpub.py:40]: This separation makes sense, let's do it.

**Review by Cito — APPROVED** [review]
Looks good to me. I would rephrase one docstring, see below.

File: hexkit/protocols/eventsub.py (hunk 7 above)
Comment (KerstenBreuer) [reply hexkit/protocols/eventsub.py:64]: Agreed.

Comment (KerstenBreuer) [reply hexkit/protocols/eventsub.py]: ✅

File: hexkit/protocols/eventpub.py (hunk 8 above)
Comment (KerstenBreuer) [reply hexkit/protocols/eventpub.py:40]: Great, thanks.
